Upgrade to Go1.18

djmitche opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

Updates everything to Go1.18.5.

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

General agent QA -- at this stage, running in various clusters and looking for issues.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

djmitche committed:
Update to Go1.18.5

 * builds from https://github.com/DataDog/datadog-agent-buildimages/pull/270
 * builds from https://github.com/DataDog/datadog-agent-builders/pull/224
 * already using the new macos build image

maycmlee reviewed (approved):
👍 for docs.

paulcacheux commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/DataDog/datadog-agent
 
-go 1.17
+go 1.18
```
You also need to update the `.go-version` at the root, with `1.18.5`

djmitche replied:
Huh, I think that got reverted somehow.

djmitche commented:
I didn't realize I had forgotten to mark this as "draft" - sorry!

djmitche commented:
It looks like the only two non-green things here were due to GitHub workers dying, not to Go1.18.  One is CodeQL, which shouldn't depend on the Go version.  The other is the MacOS tests, which have secretly been running Go1.18 for a day or two already without trouble.  At any rate, I've re-run them both.  Now it's time for review.

paulcacheux commented:
Great work thanks ! Do you plan to setup a go workspace afterwards ? https://go.dev/doc/tutorial/workspaces

djmitche commented:
> Great work thanks ! Do you plan to setup a go workspace afterwards ? https://go.dev/doc/tutorial/workspaces

I wasn't planning on doing so myself.  However, we've been discussing having each component bundle be a distinct module, in which case go workspaces might be very useful.  Are you interested in digging into that idea?

djmitche committed:
Merge branch 'main' into dustin.mitchell/asc-128

djmitche committed:
Update Go versions in go.mod's from merge

djmitche committed:
Temporarily disable trace-agent fuzz testing

This was not enabled before Go1.18 and appears to be failing.  It can be
fixed in a followup

ahmed-mez commented on pkg/trace/agent/fuzz_test.go:
```diff
@@ -3,7 +3,7 @@
 // Copyright 2016-present Datadog, Inc.
 
-//go:build go1.18
+//go:build go1.18 && disabled
```
```suggestion
//go:build go1.18 && !windows
```
As a workaround to https://github.com/golang/go/issues/52268 - the [fix](https://github.com/golang/go/commit/c3470ca83cfdbcc71ce162b00d6b9a0004a7f649) is in 1.19

djmitche committed:
Disable fuzz testing on windows

Co-authored-by: Ahmed Mezghani <38987709+ahmed-mez@users.noreply.github.com>
